locations/jclouds/src/test/java/org/apache/brooklyn/location/jclouds/JcloudsLocationReachabilityPredicateInstantiation.java

```diff
@@ -0,0 +1,147 @@
+        predicateConfig.putStringKey(CloudLocationConfig.POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE.getName() + ".key1", "val1");
+        predicateConfig.putStringKey(CloudLocationConfig.POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE.getName() + ".key2", "val2");
+        jcloudsLocation.getFirstReachableAddress(node, predicateConfig);
+        Assert.assertEquals(TEST_INIT_MAP_CONSTRUCTOR_COUNTER.get(), 1);
+    }
+
+    @Test
+    public void testInitMapConstructorWithFlags() {
+        ConfigBag predicateConfig = new ConfigBag();
+        predicateConfig.put(CloudLocationConfig.POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE_TYPE, RecordingReachabilityCheckMapAndFlagsConstructor.class);
+        List<String> hostsMatchedHolder = new ArrayList<>();
+        predicateConfig.putStringKey(CloudLocationConfig.POLL_FOR_FIRST_REACHABLE_ADDRESS_PREDICATE.getName() + ".hostsMatchedHolder", hostsMatchedHolder);
+        jcloudsLocation.getFirstReachableAddress(node, predicateConfig);
+        Assert.assertTrue(hostsMatchedHolder.contains(ALLOWED_HOST_AND_PORT.getHostText()));
+    }
+
+    // Needs to be static in order to be instantiated dynamically
+    public static class RecordingReachabilityCheck implements Predicate<HostAndPort> {
+        @Override
+        public boolean apply(@Nullable HostAndPort input) {
+            TEST_INIT_DEFAULT_CONSTRUCTOR_COUNTER.incrementAndGet();
+            return true;
+        }
+    }
+
+    public static class RecordingReachabilityCheckMapConstructor implements Predicate<HostAndPort> {
+        private Map<String, Object> flags;
+
+        public RecordingReachabilityCheckMapConstructor(Map<String, Object> flags) {
+            this.flags = flags;
+        }
+
+        @Override
+        public boolean apply(@Nullable HostAndPort input) {
+            // TODO bad practice to assert in thread which doesn't pass Assert exceptions
+            Assert.assertEquals(flags.get("key1"), "val1");
+            Assert.assertEquals(flags.get("key2"), "val2");
+            TEST_INIT_MAP_CONSTRUCTOR_COUNTER.getAndIncrement();
+            return true;
+        }
+    }
+
+    public static class RecordingReachabilityCheckMapAndFlagsConstructor implements Predicate<HostAndPort> {
+        private Map<String, Object> flags;
+
+        public RecordingReachabilityCheckMapAndFlagsConstructor(Map<String, Object> flags) {
+            this.flags = flags;
+        }
+
+        @Override
+        public boolean apply(@Nullable HostAndPort input) {
+            ((List<String>)flags.get("hostsMatchedHolder")).add(input.getHostText());
+            return true;
+        }
+    }
+
+    public static class BailOutJcloudsLocation extends JcloudsLocation {
+        public BailOutJcloudsLocation() {
+            super();
+        }
+
+        @Override
+        public String getFirstReachableAddress(NodeMetadata nodeMetada, ConfigBag setup) {
+            return super.getFirstReachableAddress(nodeMetada, setup);
+        }
+    }
+}
```

Comment:
Well done! I like how you were able to unit test the location code.
I think you should add negative tests as well - for the case where the node is not reachable - either with minimal timeouts or as integration tests.